File: general/dynamic programming/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if(root==null){
            return 0;
        }
        computeMaxPath(root);
        return maxSum;
    }

    public int computeMaxPath(TreeNode root){
        if(root==null){
            return 0;
        }

        int left=Math.max(0,computeMaxPath(root.left)); //ignore negative
        int right= Math.max(0,computeMaxPath(root.right)); 

        maxSum= Math.max(maxSum,left+right+root.val);

        return Math.max(left,right)+root.val; // check if left is bigger or right and nodes to it
    }
}

File: general/dynamic programming/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.ts
function maxPathSum(root: TreeNode | null): number {
    // We use a closure so the dfs function can update this variable
    let maxSum = -Infinity; // FIXED: Handles trees where all nodes are negative

    function dfs(node: TreeNode | null): number {
        if (node === null) return 0;

        // FIXED: Traverse first, and ignore paths that evaluate to less than 0
        let left = Math.max(0, dfs(node.left));
        let right = Math.max(0, dfs(node.right));

        // Job 1: Check if the "V-shape" path here is the new global max
        maxSum = Math.max(maxSum, node.val + left + right);

        // Job 2: Pass up ONLY a straight line (root + best single child path)
        // FIXED: We do not return left + right + node.val here
        return node.val + Math.max(left, right); 
    }

    dfs(root);
    return maxSum;
}